src/ldapauth.c

```diff
@@ -0,0 +1,956 @@
+		log_warning("could not set LDAP protocol version: %s", ldap_err2string(r));
+		ldap_unbind(*ldap);
+		return false;
+	}
+
+	ts.tv_sec = 3;
+	ts.tv_usec = 0;
+	if ((r = ldap_set_option(*ldap, LDAP_OPT_NETWORK_TIMEOUT, &ts)) != LDAP_SUCCESS) {
+		log_warning("could not set LDAP timeout: %s", ldap_err2string(r));
+		ldap_unbind(*ldap);
+		return false;
+	}
+
+	if (request->ldaptls) {
+		if ((r = ldap_start_tls_s(*ldap, NULL, NULL)) != LDAP_SUCCESS) {
+			log_warning("could not start LDAP TLS session: %s, server: %s, port: %d",
+				    ldap_err2string(r), request->ldapserver, request->ldapport);
+			ldap_unbind(*ldap);
+			return false;
+		}
+	}
+
+	return true;
+}
+/* Placeholders recognized by format_search_filter.  For now just one. */
+#define LPH_USERNAME "$username"
+#define LPH_USERNAME_LEN strlen(LPH_USERNAME)
+/*
+ * Return a newly allocated C string copied from "pattern" with all
+ * occurrences of the placeholder "$username" replaced with "user_name".
+ */
+static void format_search_filter(char *filter, int length, const char *pattern, const char *user_name)
+{
+	int cur_len = 0;
+	while ((*pattern != '\0') && (cur_len < length)) {
+		if (strncmp(pattern, LPH_USERNAME, LPH_USERNAME_LEN) == 0) {
+			cur_len += snprintf(filter + cur_len, length - cur_len, "%s", user_name);
+			pattern += LPH_USERNAME_LEN;
+		} else {
+			filter[cur_len++] = *pattern++;
+		}
+	}
+	if (cur_len >= length)
+		cur_len = length - 1;
+	filter[cur_len] = '\0';
+}
+/*
+ * Perform LDAP authentication
+ */
+static bool check_ldap_auth(struct ldap_auth_request *request)
+{
+	LDAP *ldap;
+	int r;
+	char fulluser[LDAP_LONG_LENGTH];
+
+	if (!initialize_ldap_parameters(request, request->client->ldap_parameters)) {
+		return false;
+	}
+	if ((!request->ldapserver || request->ldapserver[0] == '\0') &&
+	    (!request->ldapbasedn || request->ldapbasedn[0] == '\0')) {
+		log_warning("LDAP server not specified, and no ldapbasedn");
+		return false;
+	}
+
+	if (request->ldapport == 0) {
+		if (request->ldapscheme != NULL &&
+		    strcmp(request->ldapscheme, "ldaps") == 0)
+			request->ldapport = LDAPS_PORT;
+		else
+			request->ldapport = LDAP_PORT;
+	}
+
+	if (request->password[0] == '\0') {
+		return false;
+	}
+
+	if (InitializeLDAPConnection(request, &ldap) == false) {
+		return false;
+	}
+
+	if (request->ldapbasedn) {
+		/*
+		 * First perform an LDAP search to find the DN for the user we are
+		 * trying to log in as.
+		 */
+		char filter[LDAP_LONG_LENGTH];
+		LDAPMessage *search_message;
+		LDAPMessage *entry;
+		char *attributes[2] = {LDAP_NO_ATTRS, NULL};
+		char *dn;
+		char *c;
+		int count;
+
+		/*
+		 * Disallow any characters that we would otherwise need to escape,
+		 * since they aren't really reasonable in a username anyway. Allowing
+		 * them would make it possible to inject any kind of custom filters in
+		 * the LDAP filter.
+		 */
+		for (c = request->username; *c; c++) {
+			if (*c == '*' ||
+			    *c == '(' ||
+			    *c == ')' ||
```

Comment:
Based on your test cases I think this character list should also include `=`, `|`, `'`, `"`, `&`, and `?`. Maybe there are even more. Is there an official list of these special characters that we could follow? (and link to in the comment).

Otherwise maybe a better way is to use URI encoding (aka percent encoding) for any character that's not alphanumeric, dash or underscore

Reply:
These codes are aligned with postgres LDAP authentication.
I just searched 2 [rfc 2253](https://datatracker.ietf.org/doc/html/rfc2253) and [rfc 2254](https://datatracker.ietf.org/doc/html/rfc2254)
The username could be inside the "search filter" and "DN" in ldap, I guess the original intension according to the comments for this line of code is that disallow these special character inside usernames to avoid possible injection in the filter. This could be more strict than RCF but safer for our code.
So any username includes one of character of "*()\/" would definitely failed for LDAP authentication.

Reply:
Thank you for that info, I think this escaping is fine then.

Reply:
Thanks for the review! 
Is there anything else that needs to be modified for this PR?

Reply:
No you're good. I'll probably only merge it in ~1 month though, due to some planning for the next patch release and some time off on my side.

Reply:
Thanks again @JelteF ! This PR takes so long, if it could be accepted, I would be really happy from inside.